File: leet/0173_bst_iterator/0173_bst_iterator.js
// LeetCode 173. Binary Search Tree Iterator
// O(1) time O(h) space
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */

var BSTIterator = function(root) {
    this.stack = [];
    this.pushLeft(root);
};

/**
 * @param {TreeNode} root
 */
BSTIterator.prototype.pushLeft = function(root) {
    while (root) {
        this.stack.push(root);
        root = root.left;
    }
}

/**
 * @return {number}
 */
BSTIterator.prototype.next = function() {
    let node = this.stack.pop();
    this.pushLeft(node.right);
    return node.val;
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function() {
    return this.stack.length > 0;
};

/** 
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

File: leet/0173_bst_iterator/0173_bst_iterator.py
# LeetCode 173. Binary Search Tree Iterator
# O(1) time O(h) space
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.stack = []
        self.pushLeft(root)

    def pushLeft(self, root):
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        node = self.stack.pop()
        self.pushLeft(node.right)
        return node.val

    def hasNext(self) -> bool:
        return len(self.stack)

# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()